vowpalwabbit/metrics.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "debug_log.h"
+#include "reductions.h"
+#include "learner.h"
+#include <rapidjson/filewritestream.h>
+#include <rapidjson/writer.h>
+#include <cfloat>
+
+#include "io/logger.h"
+
+using namespace VW::config;
+using namespace VW::LEARNER;
+using namespace rapidjson;
+
+namespace logger = VW::io::logger;
+
+namespace VW
+{
+namespace metrics
+{
+struct metrics_data
+{
+  std::string out_file;
+  size_t learn_count = 0;
+  size_t predict_count = 0;
+  size_t predicted_first_option = 0;
+  size_t predicted_not_first = 0;
+};
+
+void list_to_json_file(std::string filename, std::vector<std::tuple<std::string, size_t>>& metrics)
+{
+  FILE* fp;
+
+  if (VW::file_open(&fp, filename.c_str(), "wt") == 0)
+  {
+    char writeBuffer[1024];
+    FileWriteStream os(fp, writeBuffer, sizeof(writeBuffer));
+    Writer<FileWriteStream> writer(os);
+
+    writer.StartObject();
+    for (std::tuple<std::string, size_t> m : metrics)
+    {
+      writer.Key(std::get<0>(m).c_str());
+      writer.Int64(static_cast<int32_t>(std::get<1>(m)));
+    }
+    writer.EndObject();
+
+    fclose(fp);
+  }
+  else
+  {
+    logger::errlog_warn("warning: skipping metrics. could not open file for metrics: {}", filename);
```

Comment:
Warning is implied by the log level

```suggestion
    logger::errlog_warn("skipping metrics. could not open file for metrics: {}", filename);
```